src/watch.ts

```diff
@@ -12,189 +13,196 @@ import { datadog } from './utils/datadog';
-      }
-    })
-    .on('error', (err) => {
-      sentry.report(err);
+  constructor(stateManager: StateManager, mainIndex: SearchIndex) {
+    this.stateManager = stateManager;
+    this.mainIndex = mainIndex;
+  }
+
+  /**
+   * Run watch.
+   *
+   *  --- Watch ?
+   *   Watch is "Long Polled. This mode is not paginated and the event system in CouchDB send
+   *     events as they arrive, which is super cool and reactive.
+   *   One gotcha those events arrive at the same rate wether you are watching the last seq or not.
+   *
+   *   Example:
+   *    listener A - up to date
+   *    listener B - few sequences behind.
+   *
+   *    Package C is updated.
+   *
+   *    Listener A receive update C
+   *    listener B receive update N.
+   *
+   *    Listener A is up to date again
+   *    listener B is still few sequences behind and will not receive any other event
+   *      until an other package is updated.
+   *      It will never be up to date because he receive event at the same pace
+   *      as they arrive in listener A, even if it's not the same package.
+   */
+  async run(): Promise<void> {
+    log.info('-----');
+    log.info('🚀  Watch: starting');
+    log.info('-----');
+
+    await this.stateManager.save({
+      stage: 'watch',
     });
 
-  changesConsumer.saturated(() => {
-    if (changesConsumer.length() < 5) {
-      npm.db.changesReader.resume();
-    }
-  });
+    this.changesConsumer = this.createChangeConsumer();
 
-  return new Promise((resolve) => {
-    listener.on('end', () => {
-      resolve(true);
-    });
-  });
-}
+    setInterval(async () => {
+      this.totalSequence = (await npm.db.info()).update_seq;
+    }, 5000);
 
-/**
- * Process changes in order.
- */
-async function loop(
-  mainIndex: SearchIndex,
-  change: DatabaseChangesResultItem
-): Promise<void> {
-  datadog.increment('packages');
-
-  if (!change.id) {
-    // Can happen when NPM send an empty line (for example the hearthbeat) 🤷🏻‍
-    log.error('Got a document without name', change);
-    return;
-  }
+    await this.watch();
 
-  try {
-    if (change.deleted) {
-      // Delete package directly in index
-      // Filter does not support async/await but there is no concurrency issue with this
-      throw new Error('deleted');
-    }
-    const res = await npm.getDoc(change.id, change.changes[0].rev);
+    log.info('-----');
+    log.info('🚀  Watch: done');
+    log.info('-----');
+  }
 
-    if (isFailure(res)) {
-      log.error('Got an error', res.error);
+  /**
+   * Active synchronous mode with Registry.
+   * Changes are polled with a keep-alived connection.
+   *
+   * @param stateManager - The state manager.
+   * @param mainIndex - Algolia index manager.
+   */
+  async watch(): Promise<boolean | undefined> {
+    const { seq } = await this.stateManager.get();
+    const changesConsumer = this.changesConsumer;
+    if (!changesConsumer) {
+      log.warn('Can not watch without a consumer');
       return;
```

Comment:
if we return `undefined`, is the eslint rule disabling still needed?

```suggestion
      return undefined;
```

Reply:
There is an issue for that. we could disable the rule entirely for TS though 
https://github.com/typescript-eslint/typescript-eslint/issues/1277